File: app/src/main/java/com/rainbowguo/ethexplore/viewModels/blockFragmentMode.java
package com.rainbowguo.ethexplore.viewModels;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.rainbowguo.ethexplore.Utils.TextUtils;
import com.rainbowguo.ethexplore.beans.blockInfoBean;
import com.rainbowguo.ethexplore.https.EtherScanServer;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class blockFragmentMode extends ViewModel {
    private final MutableLiveData<String> blockNumber = new MutableLiveData<>();
    private final MutableLiveData<String> date = new MutableLiveData<>();
    private final MutableLiveData<String> miner = new MutableLiveData<>();
    private final MutableLiveData<Integer> transactionsSize = new MutableLiveData<>();
    private List<String> transactionsList = new ArrayList<>();

    public MutableLiveData<String> getBlockNumber() {
        return blockNumber;
    }

    public MutableLiveData<String> getDate() {
        return date;
    }

    public MutableLiveData<String> getMiner() {
        return miner;
    }

    public List<String> getTransactionsList() {
        return transactionsList;
    }

    public MutableLiveData<Integer> getTransactionsSize() {
        return transactionsSize;
    }

    public void requestBlockData(String number){
        number = TextUtils.to16(number);
        EtherScanServer.getInstance().getBlockInfo(number, new Callback<blockInfoBean>() {
            @Override
            public void onResponse(Call<blockInfoBean> call, Response<blockInfoBean> response) {
                blockInfoBean bean = response.body();
                blockNumber.setValue(TextUtils.to10(bean.getResult().getNumber()));
                String timeStamp = TextUtils.to10(bean.getResult().getTimestamp());
                date.setValue(TextUtils.timeStampFormat(timeStamp));
                miner.setValue(bean.getResult().getMiner());
                transactionsList.addAll(bean.getResult().getTransactions());
                int listSize = transactionsList.size();
                transactionsSize.setValue(listSize);
            }

            @Override
            public void onFailure(Call<blockInfoBean> call, Throwable t) {

            }
        });
    }

}

File: app/src/main/java/com/rainbowguo/ethexplore/viewModels/blockFragmentMode.kt
package com.rainbowguo.ethexplore.viewModels

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.rainbowguo.ethexplore.Utils.TextUtils
import com.rainbowguo.ethexplore.Utils.mToast
import com.rainbowguo.ethexplore.https.HttpUtils
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import java.util.ArrayList

class blockFragmentMode : ViewModel() {
    val blockNumber = MutableStateFlow("")
    val date = MutableStateFlow("")
    val miner = MutableStateFlow("")
    val requestState = MutableStateFlow(STATE(null))
    private val transactionsList:  ArrayList<String> = ArrayList()
    private val TAG = "blockFragmentMode"
    fun getTransactionsList(): List<String> {
        return transactionsList
    }

    fun requestBlockData(number: String) {
        viewModelScope.launch(CoroutineExceptionHandler{ _, e ->
            Log.i(TAG, "requestData: $e")
            viewModelScope.launch {
                requestState.emit(STATE(false))
            }
            mToast.showToastRequestFail()
        }) {
            val bean = HttpUtils.SearchService.get_ProxyBlockInfo(TextUtils.to16(number))
            Log.i(TAG, "requestBlockData: ok ${bean.result.transactions}")
            val timeStamp = TextUtils.to10(bean.result.timestamp)
            blockNumber.value = TextUtils.to10(bean.result.number)
            date.value = TextUtils.timeStampFormat(timeStamp)
            miner.value = bean.result.miner
            transactionsList.addAll(bean.result.transactions)
            requestState.emit(STATE(true))
        }

    }

    data class STATE(val state:Boolean?)
}